lib/benchmark/linear.go

```diff
@@ -0,0 +1,77 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+)
+
+// Linear generator
+type Linear struct {
+	LowerBound          int
+	UpperBound          int
+	Step                int
+	MinimumMeasurements int
+	MinimumWindow       time.Duration
+	Threads             int
+	currentRPS          int
+	config              Config
+}
+
+// Generate advances the Generator to the next generation.
+func (lg *Linear) Generate() bool {
+	if lg.currentRPS < lg.LowerBound {
+		lg.currentRPS = lg.LowerBound
+		return true
+	}
+	lg.currentRPS += lg.Step
+	return lg.currentRPS <= lg.UpperBound
+}
+
+// GetBenchmark returns the benchmark config for the current generation.
+func (lg *Linear) GetBenchmark() (Config, error) {
+	if lg.currentRPS > lg.UpperBound {
+		return Config{}, errors.New("No more generations")
+	}
+
+	return Config{
+		MinimumWindow:       lg.MinimumWindow,
+		MinimumMeasurements: lg.MinimumMeasurements,
+		Rate:                lg.currentRPS,
+		Threads:             lg.Threads,
+		Command:             lg.config.Command,
+	}, nil
+}
+
+// Benchmark runs the benchmark of receiver type
+func (lg *Linear) Benchmark(ctx context.Context, cmd string, tc *client.TeleportClient) ([]Result, error) {
+	var result Result
+	var results []Result
+	c := strings.Split(cmd, " ")
+	lg.config.Command = c
+	sleep := false
+
+	for lg.Generate() {
+		// artificial pause between generations to allow remote state to pause
+		if sleep {
+			time.Sleep(PauseTimeBetweenBenchmarks)
+		}
+		benchmarkC, err := lg.GetBenchmark()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		result, err = benchmarkC.Benchmark(context.TODO(), tc)
```

Comment:
Why is this not using the provided context? Isn't this going to break with a misbehaving benchmark or one that takes too long?
Also, there's no special need for `sleep` - you could sleep at the end of the loop.
Is there a reason why the select loop for sleeping was not implemented as Sasha suggested?

Reply:
This is fixed here https://github.com/gravitational/teleport/pull/4643